internal/kubeadapter/provider.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2021 Crunchy Data Solutions, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package kubeadapter
+
+import (
+	"context"
+
+	"github.com/CrunchyData/crunchy-bridge-operator/internal/bridgeapi"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// kubeSecretCredentialProvider provides a LoginCred reflecting the Client's
+// current knowledge of the secret
+type KubeSecretCredentialProvider struct {
+	// Client is a reference to the kube api client
+	Client client.Client
+	// Namespace in which the secret lives
+	Namespace string
+	// Object name of the secret
+	Name string
+	// KeyField identifies the secret's subkey to map to LoginCred.Key
+	KeyField string
+	// SecretField identifies the secret's subkey to map to LoginCred.Secret
+	SecretField string
+}
+
+// TODO: untested
+func (ks *KubeSecretCredentialProvider) ProvideCredential() bridgeapi.LoginCred {
+	formedCred := bridgeapi.LoginCred{}
+	secret := &corev1.Secret{}
+	selector := client.ObjectKey{
+		Namespace: ks.Namespace,
+		Name:      ks.Name,
+	}
+
+	if err := ks.Client.Get(context.Background(), selector, secret); err != nil {
```

Comment:
Adding an error return would break the CredentialProvider interface and as with prior comments this is invoked asynchronously.

I agree that a correct demonstration is in order and will be in with the provider operator updates. The short version is that this adapter allows the credentals in the kube secret to be pulled each time the login is refreshed, so its possible to update the kube secret and thus the API login without having to restart the process.

In the draft you shared, you used the CredentialProvider interface to retrieve a LoginCred (which also implements CredentialProvider) instead of using the KubeSecretCredentialProvider directly as satisfying the CredentialProvider interface. This is due in part to your change breaking the interface and making it unusable this way.

Reply:
> Adding an error return would break the CredentialProvider interface and as with prior comments this is invoked asynchronously.
> 
> I agree that a correct demonstration is in order and will be in with the provider operator updates. The short version is that this adapter allows the credentals in the kube secret to be pulled each time the login is refreshed, so it's possible to update the kube secret and thus the API login without having to restart the process.

So if not returning the error, what should be an error if a secret not found, or what operator logs show about the exact error?
> 
> In the draft you shared, you correctly built a `KubeSecretCredentialProvider` but used the `CredentialProvider` interface to retrieve a `LoginCred` (which also implements `CredentialProvider`) instead of using the `KubeSecretCredentialProvider` directly as satisfying the `CredentialProvider` interface. This is due in part to your change breaking the interface and making it unusable this way.

Reply:
> So if not returning the error, what should be an error if a secret not found, or what operator logs show about the exact error?
Right now, the client data retrieval functions would return an error "unexpected response status from API". The logs, if connected would show the unexpected status code being something like 403 (Forbidden). This is less clear than it should be, hence #12
There would also be errors logged to the package logger from the login function showing a failure to unmarshal the token response as the body would not have a valid token.
 
This happens because a failure to obtain correct credentials results in a zero-value LoginCred for that call to ProvideCredential(), which will fail to obtain a permissible bearer token, causing future attempts to contact the API to fail (unless there is a valid token cached that has not yet expired).